drivers/network/dd/dc21x4/init.c

```diff
@@ -0,0 +1,1147 @@
+    {
+        Status = NdisMRegisterIoPortRange((PVOID*)&Adapter->IoBase,
+                                          Adapter->AdapterHandle,
+                                          Adapter->IoBaseAddress.LowPart,
+                                          DC_IO_LENGTH);
+    }
+    else
+    {
+        Status = NdisMMapIoSpace((PVOID*)&Adapter->IoBase,
+                                 Adapter->AdapterHandle,
+                                 Adapter->IoBaseAddress,
+                                 DC_IO_LENGTH);
+
+        Adapter->Flags |= DC_IO_MAPPED;
+    }
+    if (Status != NDIS_STATUS_SUCCESS)
+        goto Cleanup;
+
+Cleanup:
+    NdisFreeMemory(AssignedResources, ResourceListSize, 0);
+
+    return Status;
+}
+
+static
+CODE_SEG("PAGE")
+ULONG
+DcGetBusModeParameters(
+    _In_ PDC21X4_ADAPTER Adapter,
+    _In_ PPCI_COMMON_CONFIG PciData)
+{
+    ULONG DefaultMode, NewMode;
+
+    PAGED_CODE();
+
+    /* TODO: Other architectures than x86 */
+    DefaultMode = DC_BUS_MODE_CACHE_ALIGNMENT_16 | DC_BUS_MODE_BURST_LENGTH_NO_LIMIT;
+
+    if (!(Adapter->Features & DC_ENABLE_PCI_COMMANDS))
+        return DefaultMode;
+
+    INFO("PCI Cache Line Size %u\n", PciData->CacheLineSize * 4);
+    INFO("PCI Command %04lx\n", PciData->Command);
+
+    /* Use the cache line size if it was set up by firmware */
+    switch (PciData->CacheLineSize)
+    {
+        case 8:
+            NewMode = DC_BUS_MODE_CACHE_ALIGNMENT_8 | DC_BUS_MODE_BURST_LENGTH_8;
+            break;
+        case 16:
+            NewMode = DC_BUS_MODE_CACHE_ALIGNMENT_16 | DC_BUS_MODE_BURST_LENGTH_16;
+            break;
+        case 32:
+            NewMode = DC_BUS_MODE_CACHE_ALIGNMENT_32 | DC_BUS_MODE_BURST_LENGTH_32;
+            break;
+
+        default:
+            return DefaultMode;
+    }
+
+    /* Enable one of those commands */
+    if (PciData->Command & PCI_ENABLE_WRITE_AND_INVALIDATE)
+    {
+        NewMode |= DC_BUS_MODE_WRITE_INVALIDATE;
+    }
+    else
+    {
+        NewMode |= DC_BUS_MODE_READ_LINE;
+    }
+
+    return NewMode;
+}
+
+static
+CODE_SEG("PAGE")
+NDIS_STATUS
+DcRecognizeHardware(
+    _In_ PDC21X4_ADAPTER Adapter)
+{
+    UCHAR Buffer[RTL_SIZEOF_THROUGH_FIELD(PCI_COMMON_CONFIG, CacheLineSize)];
+    PPCI_COMMON_CONFIG PciConfig = (PPCI_COMMON_CONFIG)Buffer; // Partial PCI header
+    PNDIS_TIMER_FUNCTION MediaMonitorRoutine;
+    ULONG Bytes;
+
+    PAGED_CODE();
+
+    Bytes = NdisReadPciSlotInformation(Adapter->AdapterHandle,
+                                       0,
+                                       FIELD_OFFSET(PCI_COMMON_CONFIG, VendorID),
+                                       Buffer,
+                                       sizeof(Buffer));
+    if (Bytes != sizeof(Buffer))
+        return NDIS_STATUS_FAILURE;
+
+    Adapter->DeviceId = PciConfig->DeviceID;
+    Adapter->RevisionId = PciConfig->RevisionID;
+
+    switch ((PciConfig->DeviceID << 16) | PciConfig->VendorID)
+    {
+        case 0x00021011: /* 21040 */
```

Comment:
You can add constants in headers for known PCI device identifiers.

Reply:
Done.